libafl_qemu/src/qemu/qemu_config.rs

```diff
@@ -0,0 +1,378 @@
+pub enum RamSize {
+    MB(u32),
+    GB(u32),
+}
+
+impl Display for RamSize {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            RamSize::MB(mb) => write!(f, "-m {mb}M"),
+            RamSize::GB(gb) => write!(f, "-m {gb}G"),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct SmpCpus {
+    pub cpus: u32,
+}
+
+impl Display for SmpCpus {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "-smp {}", self.cpus)
+    }
+}
+
+#[derive(Debug, Clone, strum_macros::Display)]
+pub enum VgaPci {
+    #[strum(serialize = "-device VGA")]
+    ENABLE,
+    #[strum(serialize = "")]
+    DISABLE,
+}
+
+impl From<bool> for VgaPci {
+    fn from(vga_pci: bool) -> Self {
+        if vga_pci {
+            VgaPci::ENABLE
+        } else {
+            VgaPci::DISABLE
+        }
+    }
+}
+
+#[cfg(emulation_mode = "usermode")]
+#[derive(Debug, Clone)]
+pub struct Program {
+    path: PathBuf,
+}
+
+#[cfg(emulation_mode = "usermode")]
+impl Display for Program {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.path.to_str().unwrap())
+    }
+}
+
+#[cfg(emulation_mode = "usermode")]
+impl<R: AsRef<Path>> From<R> for Program {
+    fn from(path: R) -> Self {
+        Self {
+            path: path.as_ref().to_path_buf(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, libafl_derive::Display, TypedBuilder)]
+#[builder(build_method(into = Result<Qemu, QemuInitError>), builder_method(vis = "pub(crate)"))]
+pub struct QemuConfig {
+    #[cfg(emulation_mode = "systemmode")]
+    #[builder(default, setter(strip_option))]
+    accelerator: Option<Accelerator>,
+    #[builder(default, setter(strip_option, into))]
+    bios: Option<Bios>,
+    #[builder(default)]
+    drives: Vec<Drive>,
+    #[builder(default, setter(strip_option, into))]
+    kernel: Option<Kernel>,
+    #[builder(default, setter(strip_option, into))]
+    load_vm: Option<LoadVM>,
+    #[builder(default, setter(strip_option, into))]
+    machine: Option<Machine>,
+    #[builder(default, setter(strip_option))]
+    monitor: Option<Monitor>,
+    #[builder(default, setter(strip_option, into))]
+    no_graphic: Option<NoGraphic>,
+    #[builder(default, setter(strip_option))]
+    ram_size: Option<RamSize>,
+    #[builder(default, setter(strip_option))]
+    serial: Option<Serial>,
+    #[builder(default, setter(strip_option))]
+    smp_cpus: Option<SmpCpus>,
+    #[builder(default, setter(strip_option, into))]
+    snapshot: Option<Snapshot>,
+    #[builder(default, setter(strip_option, into))]
+    vga_pci: Option<VgaPci>,
+    #[builder(default, setter(strip_option, into))]
+    start_cpu: Option<StartCPU>,
+    #[cfg(emulation_mode = "usermode")]
+    #[builder(setter(into))]
+    program: Program,
+} // Adding something here? Please leave Program as the last field
+
+impl From<QemuConfig> for Result<Qemu, QemuInitError> {
+    fn from(config: QemuConfig) -> Self {
```

Comment:
Not sure a simple `into()` should mutate global state, maybe consider making this a "normal" function with documentation

Reply:
This is required to make the API look like a normal builder `qemu.builder().foo(bar).build()` and still use TypedBuilder even with our "not standard" use case where `Qemu` doesn't store the config.
Under the hood TypedBuilder generates a builder for `QemuConfig`. `QemuConfig.into()` is used in `QemuConfigBuilder.build()`
If you have any better idea how to solve this I'd be happy to change this janky thing 😄
(Unfortunately visibility of traits impl cannot be set, but still this impl should be safe from misusages)

Reply:
Maybe move init behind the global access to make sure we don't init twice(?)